File: java/src/main/java/net/imagini/aim/partition/AimPartitionServerQuerySession.java
package net.imagini.aim.partition;

import java.io.EOFException;
import java.io.IOException;
import java.util.ArrayList;
import java.util.Queue;
import java.util.concurrent.ExecutionException;

import net.imagini.aim.cluster.AimQuery;
import net.imagini.aim.segment.AimFilter;
import net.imagini.aim.tools.Pipe;
import net.imagini.aim.tools.Tokenizer;
import net.imagini.aim.types.AimDataType;
import net.imagini.aim.types.AimSchema;
import net.imagini.aim.types.AimType;

import org.apache.commons.lang3.StringUtils;

public class AimPartitionServerQuerySession extends Thread {

    private Pipe pipe;
    private AimPartition table;

    public AimPartitionServerQuerySession(AimPartition table, Pipe pipe) throws IOException {
        this.pipe = pipe;
        this.table = table;
    }

    @Override
    public void run() {
        try {
            AimSchema schema = table.schema;
            AimQuery query = new AimQuery(table);
            Integer range = null;
            AimFilter filter = null;
            while(true) {
                String command;
                Queue<String> cmd;
                String input = pipe.read();
                try {
                    cmd = Tokenizer.tokenize(input);
                    command = cmd.poll().toUpperCase();
                    switch(command.toUpperCase()) {

                        default: break;

                        case "STATS": handleStats(cmd); break;

                        case "ALL": range = null; filter = null; break;
                        case "LAST": range = table.getNumSegments()-1; filter = null; break;
                        case "RANGE": handleRangeQuery(cmd,query); filter = null; break;
                        case "FILTER": 
                            if (cmd.size()>0) {
                                filter = AimFilter.fromTokenQueue(schema, cmd); //range, query,  
                            }
                            executeCount(range, query, filter);
                            break;

                        case "SELECT": 
                            if (cmd.size()>0) {
                                schema = table.schema.subset(new ArrayList<String>(cmd));
                            }
                            executeSelect(range, query, schema, filter, true); 
                            break;
                    }
                    pipe.write(false);
                    pipe.flush();
                } catch (Exception e) {
                    try {
                        pipe.write(true);
                        pipe.write("ERROR");
                        pipe.write(exceptionAsString(e));
                        pipe.write(false);
                        pipe.flush();
                    } catch (Exception e1) {
                        System.out.println("Could not send error response to the clien: ");
                        e.printStackTrace();
                    }
                }
            }
        } catch (IOException e) {
            try {
                pipe.close();
            } catch (IOException e1) {}
            return;
        }
    }

    private void handleStats(Queue<String> cmd) throws IOException {
        pipe.write(true);
        pipe.write("STATS");
        pipe.write(table.schema.toString());
        pipe.write(table.getCount());
        pipe.write(table.getNumSegments());
        pipe.write(table.getSize());
        pipe.write(table.getOriginalSize());
    }

    private AimQuery handleRangeQuery(Queue<String> cmd, AimQuery query) {
        // TODO Auto-generated method stub
        return null;
    }

    private void executeCount(
        Integer range, 
        AimQuery query, 
        AimFilter filter
    ) throws IOException, ExecutionException {
        long t = System.currentTimeMillis();
        t = (System.currentTimeMillis()-t);
        query.range(range);
        pipe.write(true);
        pipe.write("COUNT");
        Long count = query.count(filter);
        pipe.write(filter == null ? null : filter.toString());
        pipe.write(count);
        pipe.write((long)table.getCount());
        pipe.flush();
    }

    private void executeSelect(
        Integer range, 
        AimQuery query, 
        AimSchema schema, 
        AimFilter filter,
        boolean fetch
    ) throws IOException {
        query.range(range);
        Pipe result = query.select(filter, schema.names());
        pipe.write(true);
        pipe.write("RESULT");
        pipe.write(schema.toString());
        pipe.write(filter == null ? null : filter.toString());

        long count = 0; 
        try {
            boolean written;
            while(true) {
                written = false;
                for(AimType type: schema.fields()) {
                    AimDataType dataType = type.getDataType();
                    if (!fetch) {
                        result.skip(dataType);
                    } else {
                        byte[] val = result.read(dataType);
                        if (!written) pipe.write(written = true);
                        pipe.write(dataType.getDataType(), val);
                    }

                }
                count++;
            }
        } catch (Exception e) {
            pipe.write(false);
            pipe.write((long)count);
            pipe.write((long)table.getCount());
            pipe.write((e instanceof EOFException ? "" : exceptionAsString(e))); //success flag
            pipe.flush();
        }

    }


    public static String exceptionAsString(Exception e) {
        String[] trace = new String[e.getStackTrace().length];
        int i = 0; for(StackTraceElement t:e.getStackTrace()) trace[i++] = t.toString();
        return e.toString() + "\n"  + StringUtils.join(trace,"\n");
    }

}

File: java/src/main/java/net/imagini/aim/partition/AimPartitionServerQuerySession.scala
package net.imagini.aim.partition

import net.imagini.aim.tools.Pipe
import java.io.IOException
import net.imagini.aim.tools.AimFilter
import java.util.Queue
import net.imagini.aim.tools.Tokenizer
import java.util.ArrayList
import java.io.InputStream
import net.imagini.aim.tools.PipeUtils
import java.io.EOFException

class AimPartitionServerQuerySession(val partition: AimPartition, val pipe: Pipe) extends Thread {

  var filter: AimFilter = null

  def exceptionAsString(e: Throwable): String = e.getMessage + e.getStackTrace.map(trace ⇒ trace.toString).foldLeft("\n")(_ + _ + "\n")

  override def run = {
    try {
      while (true) {
        val input: String = pipe.read
        try {
          val cmd = Tokenizer.tokenize(input)
          val command = cmd.poll.toUpperCase
          command.toUpperCase match {
            case "STATS"  ⇒ handleStats(cmd)
            case "FILTER" ⇒ handleFilter(cmd)
            case "SELECT" ⇒ handleSelectRequest(cmd)
            case _        ⇒ {}
          }
          pipe.write(false)
          pipe.flush()
        } catch {
          case e: Throwable ⇒ try {
            pipe.write(true);
            pipe.write("ERROR");
            pipe.write(exceptionAsString(e));
            pipe.write(false);
            pipe.flush();
          } catch {
            case e1: Throwable ⇒ {
              e.printStackTrace();
            }
          }
        }
      }
    } catch {
      case e: IOException ⇒ {
        try pipe.close catch { case e: IOException ⇒ e.printStackTrace }
      }
    }
  }

  private def handleFilter(cmd: Queue[String]) = {
    if (cmd.size > 0) {
      filter = AimFilter.fromTokenQueue(partition.schema, cmd)
    }
    pipe.write(true);
    pipe.write("COUNT");
    val count = partition.count(filter)
    pipe.write(if (filter == null) null else filter.toString)
    pipe.write(count);
    pipe.write(partition.getCount)
    pipe.flush();

  }

  private def handleStats(cmd: Queue[String]) = {
    pipe.write(true)
    pipe.write("STATS")
    pipe.write(partition.schema.toString)
    pipe.write(partition.getCount)
    pipe.write(partition.getNumSegments)
    pipe.write(partition.getCompressedSize)
    pipe.write(partition.getUncompressedSize)
  }

  private def handleSelectRequest(cmd: Queue[String]) = {
    val schema = if (cmd.size > 0) partition.schema.subset(new ArrayList(cmd)) else partition.schema
    val result: InputStream = partition.select(filter, schema.names);
    pipe.write(true)
    pipe.write("RESULT")
    pipe.write(schema.toString)
    pipe.write(if (filter == null) null else filter.toString)
    var count: Long = 0
    try {
      while (true) {
        var written = false;
        for (t ← schema.fields) {
          val dataType = t.getDataType
          val value: Array[Byte] = PipeUtils.read(result, dataType)
          if (!written) {
            written = true
            pipe.write(written)
          }
          pipe.write(dataType.getDataType, value)
        }
        count += 1
      }
    } catch {
      case e: Throwable ⇒ {
        pipe.write(false)
        pipe.write(count)
        pipe.write(partition.getCount)
        pipe.write(if (e.isInstanceOf[EOFException]) "" else exceptionAsString(e)) //success flag
        pipe.flush
      }
    }

  }
}